### PR TITLE
Implements a blacklist for problematic contractor dropoff points

### PR DESCRIPTION
**Pull request by Cenrus**
## About The Pull Request

Blacklists the following areas from being valid contract dropoff locations:

External solars. Kills the target from lack of pressure and has no plating for the pod to land at.
Toxins test site, for being in space.
AI Upload, core and satellite for being excruciatingly hard and protected by turrets
Shuttles.

## Why It's Good For The Game

People dying from external dropoff locations isn't good, the point of contracting is that the victim is supposed to be sent up and return alive.

## Changelog
:cl:
add: Blacklisted solars, toxins test site, shuttles, and AI Upload/Sat from being contract dropoff locations.
/:cl:

### CODE REVIEW

File: code/game/gamemodes/objective.dm [hunk 1]
```diff
@@ -953,13 +953,17 @@ GLOBAL_LIST_EMPTY(possible_items_special)
 	var/payout = 0
 	var/payout_bonus = 0
 	var/area/dropoff = null
-
+	var/static/list/area_blacklist = typecacheof(list(/area/ai_monitored/turret_protected,
+														/area/solar/,
+														/area/science/test_area/,
+														/area/shuttle/))
```
Comment (PowerfulBacon) [inline code/game/gamemodes/objective.dm]:
```suggestion
														/area/solar,
														/area/science/test_area,
														/area/shuttle))
```

**Comment by eeSPee** [comment]
I'd also like to add, head's offices and anything sec.

**Comment by PowerfulBacon** [comment]
hold on this can literally be custom player areas made in space can't it?

**Comment by Archanial** [comment]
Why not use `GLOB.the_station_areas` instead of `sortedAreas`?
Or you can change it completely to use `get_safe_random_station_turf()`

File: code/game/gamemodes/objective.dm [hunk 2]
```diff
@@ -953,13 +953,17 @@ GLOBAL_LIST_EMPTY(possible_items_special)
...
+														/area/science/test_area,
+														/area/shuttle))
+	
 // Generate a random valid area on the station that the dropoff will happen.
 /datum/objective/contract/proc/generate_dropoff()
 	var/found = FALSE
 	while (!found)
 		var/area/dropoff_area = pick(GLOB.sortedAreas)
-		if(dropoff_area && is_station_level(dropoff_area.z) && !dropoff_area.outdoors)
+		if(dropoff_area && is_station_level(dropoff_area.z) && !dropoff_area.outdoors && !is_type_in_typecache(dropoff_area, area_blacklist))
```
Comment (utilbird) [inline code/game/gamemodes/objective.dm]:
```suggestion
		if(dropoff_area && is_station_level(dropoff_area.z) && !dropoff_area.outdoors && !area_blacklist[dropoff_area.type]
```

**Comment by Cenrus** [comment]
Not blacklisting sec/head offices from being contract targets. You can earn well over 20 TC (+ the equipment you get in the starter kit) by doing contracts, and secure dropoff locations are part of the challenge of contracting.

**Comment by PowerfulBacon** [comment]
Revert elons review, that doesnt account for subtypes of areas. Also it doesn't compile

**Comment by heepox** [comment]
No more solars or AI crap? Thank god. Now i dont have to maxcap AI inorder to contract.

**Comment by TheFakeElon** [comment]
> Revert elons review, that doesnt account for subtypes of areas. Also it doesn't compile

It does account for subtypes. This is the cleaner way to do it.

I might have typo'd and broke it though 😅

**Comment by utilbird** [comment]
> Revert elons review, that doesnt account for subtypes of areas. Also it doesn't compile

*prettyyy* sure It does account for subtypes. This is the more efficient and cleaner way to do it. Don't know why it didn't compile though, I probably typo'd somewhere 😅

File: code/game/gamemodes/objective.dm (hunk 2 above)
Comment (utilbird) [inline code/game/gamemodes/objective.dm]: OH, it was the missing closing bracket
```suggestion
		if(dropoff_area && is_station_level(dropoff_area.z) && !dropoff_area.outdoors && !area_blacklist[dropoff_area.type])
```

**Comment by ike709** [comment]
There is in fact that helper proc for getting a safe station turf.

**Comment by Archanial** [comment]
> There is in fact that helper proc for getting a safe station turf.

Yes, get_safe_random_station_turf()

**Review by francinum — CHANGES_REQUESTED** [review]
We have a proper check for this already, `area.valid_territory`

If you want to blacklist anything more specific, please generalize it as an option in map jsons than a hardcoded list.

**Comment by Sinestia** [comment]
> 
> 
> I'd also like to add, head's offices and anything sec.

Heads and sec aren't that bad, and are weird which is good

Main thing is that people shouldn't be being thrown to their deaths

**Comment by PowerfulBacon** [comment]
stale, reopen when changes applied

**Pull request closed without merging**